security/pkg/nodeagent/caclient/providers/citadel/client.go

```diff
@@ -125,6 +132,16 @@ func (c *citadelClient) getTLSDialOption() (grpc.DialOption, error) {
 	config := tls.Config{}
 	config.RootCAs = certPool
 
+	if ProvCert != "" {
+		// Load the certificate from disk
+		certificate, err := tls.LoadX509KeyPair(ProvCert+"/cert-chain.pem", ProvCert+"/key.pem")
```

Comment:
This is run one time at startup, are we ok with this breaking once the cert expires? Are these always long lived certs?

Reply:
Don't we write the certs back ? 

In general we used longer-lived certs for VMs, and after each refresh write them back to extend the life. I haven't tested this part - I think using a 6-month cert is not that bad for 1.5 (given that alternative is to stick with 1.4)

Reply:
Also - the use case for 'provisioning certs' is that a separate tool - ansible, puppet, etc - takes care of installing certs on the machine - similar with how our google laptops (and prod) works. 
The 'hack' we used to write back the certs may still work, but I'm not sure it was ever a good practice.

Reply:
Even if we write them back, I think this code is just run one time at startup and not changed even if the underlying file changes? maybe reading thiswrong

Reply:
And you are right... The code is indeed not reloading - I just copied what the VM node agent was using for initialization ( node_agent/main -> platform/NewClient -> nafactory.go ).
I guess it was working because we provision long certs.

Since it's not a regression - and I suspect the main use case will remain with long-lived provisioning certs - can we handle the rotation in separate PR or 1.6 ? One of the main benefits of this PR is that it allows use of puppet/ansible/etc to provision longer lived certs - but still use istiod to get short-lived certs with extra info for workload-to-workload. 

The code is pretty complex - and I'm a bit scared to make bigger changes in 1.5 branch.

Reply:
before the cert expires, envoy will send a SDS request to request a new cert, right? Then we will using the client to send the CSR and get the cert updated.

Reply:
Envoy doesn't send an SDS request when the cert is about to expire -- I thought it did so I may have spread some misinformation, but it turns out we are just watching this in istio-agent

Reply:
This has nothing to do with Envoy - pilot agent creates a grpc client for the ISTIO_CA ( istiod or external ), and initializes it at startup with the certificates. That's what it did for VMs before. 

It could check the cert and reload it - before it expires, no need to watch files. Since istio-agent is watching ( and I haven't reviewed that code - not sure what it would watch except the date in the cert ) - we could reuse some code, but it's a bit messy.